mbedtls/src/ssl/async_io.rs

```diff
@@ -15,29 +15,63 @@ use crate::{
         io::{IoCallback, IoCallbackUnsafe},
     },
 };
+use async_trait::async_trait;
 use std::{
     future::Future,
     io::{Error as IoError, ErrorKind as IoErrorKind, Result as IoResult},
     pin::Pin,
+    result::Result as StdResult,
     task::{Context as TaskContext, Poll},
 };
-use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    macros::support::poll_fn,
+    net::UdpSocket,
+};
 
 /// Marker type for an IO implementation that implements both
 /// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
 pub enum AsyncStream {}
 
-// TODO: Add enum `AnyAsyncIo` as marker type for an IO implementation that
-// doesn't implement `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+/// Marker type for an IO implementation that
+/// doesn't implement [`tokio::io::AsyncRead`] and [`tokio::io::AsyncWrite`].
+pub enum AnyAsyncIo {}
+
+/// Async read or write bytes or packets.
+///
+/// Implementors represent a socket or file descriptor that can be read from or
+/// written to.
+///
+/// By implementing [`AsyncIo`] you can wrap any type of
+/// [`AsyncIo`] with [`Context<T>::establish_async`] to protect that
+/// communication channel with (D)TLS. That [`Context`] then also implements
+/// [`AsyncIo`] so you can use it interchangeably.
+///
+/// If you are using byte streams and are using [`tokio::io`], you don't need
+/// this trait and can rely on [`tokio::io::AsyncRead`] and
+/// [`tokio::io::AsyncWrite`] instead.
+#[async_trait]
+pub trait AsyncIo {
+    async fn recv(&mut self, buf: &mut [u8]) -> IoResult<usize>;
+    async fn send(&mut self, buf: &[u8]) -> IoResult<usize>;
+}
+
+#[async_trait]
+impl<T: AsyncIo + Send> AsyncIo for Context<T> {
+    async fn recv(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        self.io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?
+            .recv(buf)
+            .await
+    }
 
-// TODO: Add `AsyncIo` trait for async IO that that doesn't implement
-// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`. For example:
-//     pub trait AsyncIo {
-//        async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
-//         async fn send(&mut self, buf: &[u8]) -> Result<usize>;
-//     }
-// Could implement by using `async-trait` crate or
-// #![feature(async_fn_in_trait)] or Associated Types
+    async fn send(&mut self, buf: &[u8]) -> IoResult<usize> {
+        self.io_mut()
```

Comment:
This doesn't seem correct? You need to call mbedtls_ssl_write here. Same issue for recv. I think you can avoid async-trait and some headache if you model this with `fn poll_recv` and `fn poll_send` instead of `async fn`.